default/AI/ResourcesAI.py

```diff
@@ -211,30 +381,227 @@ def assess_protection_focus(pid):
+            # calculate factor F at which II + F * IR == RI + F * RR =====> F = ( II-RI ) / (RR-IR)
+            thisFactor = (II - RI) / max(0.01, RR - IR)  # don't let denominator be zero for planets where focus doesn't change RP
+            planet = pinfo.planet
+            if adj_round == 2:  # take research at planets with very cheap research
+                if (maxi_ratio < priorityRatio) and (curTargetRP < priorityRatio * ctPP0) and (thisFactor <= 1.0):
+                    curTargetPP += RI
+                    curTargetRP += RR
+                    focus_manager.bake_future_focus(pid, RFocus, False)
+                continue
+            if adj_round == 3:  # take research at planets where can do reasonable balance
+                if has_force or (foAI.foAIstate.aggression < fo.aggression.aggressive) or (curTargetRP >= priorityRatio * ctPP0):
+                    continue
+                pop = planet.currentMeterValue(fo.meterType.population)
+                t_pop = planet.currentMeterValue(fo.meterType.targetPopulation)
+                # if AI is aggressive+, and this planet in range where temporary Research focus can get an additional RP at cost of 1 PP, and still need some RP, then do it
+                if pop < t_pop - 5:
+                    continue
+                if (CI > II + 8) or (((RR > II) or ((RR - CR) >= 1 + 2 * research_penalty)) and ((RR - IR) >= 3) and ((CR - IR) >= 0.7 * ((II - CI) * (1 + 0.1 * research_penalty)))):
+                    curTargetPP += CI - 1 - research_penalty
+                    curTargetRP += CR + 1
+                    focus_manager.bake_future_focus(pid, RFocus, False)
+                continue
+            if adj_round == 4: #assume default IFocus
+                curTargetPP += II  # icurTargets initially calculated by Industry focus, which will be our default focus
+                curTargetRP += IR
+                ratios.append((thisFactor, pid, pinfo))
+
+    ratios.sort()
+    printedHeader = False
+    gotAlgo = tech_is_complete("LRN_ALGO_ELEGANCE")
+    for ratio, pid, pinfo in ratios:
+        do_research = False  # (focus_manager.new_foci[pid]==RFocus)
+        if (priorityRatio < (curTargetRP / (curTargetPP + 0.0001))) and not do_research:  # we have enough RP
+            if ratio < 1.1 and foAI.foAIstate.aggression > fo.aggression.cautious:  # but wait, RP is still super cheap relative to PP, maybe will take more RP
+                if priorityRatio < 1.5 * (curTargetRP / (curTargetPP + 0.0001)):  # yeah, really a glut of RP, stop taking RP
+                    break
+            else:  # RP not super cheap & we have enough, stop taking it
+                break
+        II, IR = pinfo.possible_output[IFocus]
+        RI, RR = pinfo.possible_output[RFocus]
+        # if focus_manager.current_focus[pid] == MFocus:
+        # II = max( II, focus_manager.possible_output[MFocus][0] )
+        if (not do_research and (
```

Comment:
Commit [0e8bd8b676f45b 'Fix overly indented code in set_planet_industry_and_foci' ](https://github.com/freeorion/freeorion/pull/580/commits/0e8bd8b676f45b05d084149d841880f1bf5f4d01) fixes the bug.

Previously, the code to prevent planets satisfying the conditions now at line 568-573 was indented to the depth inside the `if` of the line  574 `if not printedHeader`.  It was only executed when the header was printed, instead of for each planet that satisfied the entry conditions.

The important (non reporting) line is the `continue` on line 583, which needs to run to skip the next planet.  In the buggy code it was falling through and having its focus set th RFocus.